core/primitives/src/receipt.rs

```diff
@@ -90,6 +91,73 @@ pub enum Receipt {
     V1(ReceiptV1),
 }
 
+/// A receipt that is stored in the state with added metadata. A receipt may be
+/// stored in the state as a delayed receipt, buffered receipt or a promise
+/// yield receipt. The metadata contains additional information about receipt
+///
+/// Please note that the StateStoredReceipt implements custom serialization and
+/// deserialization. Please see the comment on [ReceiptOrStateStoredReceipt]
+/// for more details.
+///
+/// This struct is versioned so that it can be enhanced in the future.
+#[derive(PartialEq, Eq, Debug)]
+pub enum StateStoredReceipt {
+    V0(StateStoredReceiptV0),
+}
+
+#[derive(BorshDeserialize, BorshSerialize, PartialEq, Eq, Debug)]
+pub struct StateStoredReceiptV0 {
+    /// The receipt.
+    pub receipt: Receipt,
+    pub metadata: StateStoredReceiptMetadata,
+}
+
+#[derive(BorshDeserialize, BorshSerialize, PartialEq, Eq, Debug)]
+pub struct StateStoredReceiptMetadata {
+    pub gas: Gas,
+    pub size: u64,
+}
+
```

Comment:
The names `StateStoredReceipt` and `ReceiptOrStateStoredReceipt` are descriptive, but a bit unwieldy and not very future proof. Maybe we could reorganize things to be more generic.
Something like:
```rust
// Renamed
// `ReceiptOrStateStoredReceipt` -> `TrieQueueItem`
// `StateStoredReceipt` -> `ReceiptWithMetadata`
enum TrieQueueItem {
    Receipt(Receipt),
    ReceiptWithMetadata(Receipt, StateStoredReceiptMetadata),
}
```

This would allow to add other things in the future, for example for `C2` we're gonna need to keep a group of receipts in the qeue, this could be another variant of the `TrieQueueItem` enum.

Reply:
I dislike it a bit because it's used for both queue and buffers so TrieQueue may be a bit misleading. On the other hand StateStoredItem is too generic.

Please keep in mind that it's the `StateStoredReceipt` that is versioned while the `ReceiptOrStateStoredReceipt` is merely a helper for the migration from `R` to `SSR`. 

let me think about it a bit more while I address the other comments

Reply:
We discussed things offline and decided to keep things as is, the other variants needed for `C2` can be represented as variants of `StorageStoredReceipt`.